Add subscription model

0xkarmacoma opened the pull request.

beskay committed:
Add Subscription Feature (and more)

- include constants MAX_REWARD and MAX_EXP
- add EpochInfo struct and epochs array
- add Subscribed event
- add more view functions
- reduce reward if total reward of epoch exceeds 10k OP
- unit tests

beskay committed:
fix typo

beskay committed:
include explanation

0xkarmacoma committed:
Combine addressToEXP and addressClaimed

Combine them into a single `epochToSubscribedEXP` mapping. The meaning of this mapping is "account's EXP balance eligible for rewards this epoch":
- it starts at 0
- when `account` subscribes, it stores the EXP balance
- when `account` claims (at epoch + 1), it resets the EXP balance to 0

This means that not subscribing and claiming look the same, but saves a significant amount on storage

0xkarmacoma committed:
Claiming automatically resubscribes

+ possible to resubscribe multiple times in the same epoch (in case of balance changes)

0xkarmacoma committed:
Update gas snapshot

Test result: ok. 14 passed; 0 failed; finished in 41.46ms
testClaimBeforeStart() (gas: 0 (0.000%))
testExtendClaim() (gas: 0 (0.000%))
testExtendClaimAuth() (gas: 0 (0.000%))
testSubscribeBeforeStart() (gas: -45 (-0.288%))
testReducedReward() (gas: -8878 (-0.695%))
testNonExpOwnerCanNotClaim() (gas: -488 (-2.034%))
testDoubleClaim() (gas: 4071 (2.731%))
testFuzzClaiming(uint256) (gas: 5058 (2.792%))
testEXPLimit() (gas: 5041 (2.838%))
testNonExpOwnerCanNotSubscribe() (gas: -2286 (-9.709%))
testExpOwnerCanClaim() (gas: -33827 (-14.321%))
testMaxClaimPeriod() (gas: -227677 (-35.617%))

beskay committed:
Fix subscribe vulnerability

- check if account already called subscribe, and if so, only add additional EXP tokens to `totalEXP`
- removed line `epochs[epochNum] = epoch` since we already access storage via `EpochInfo storage epoch`
- increased factor from 10**15 to 10**18 for more precision

beskay committed:
Update gas snapshot

beskay committed:
Refactored tests

mariobyn commented on src/OPTokenClaim.sol:
```diff
@@ -54,62 +70,139 @@ contract OPTokenClaim is Ownable {
+        emit OPClaimed(account, lastEpochNum, OPReward);
 
-        emit OPClaimed(account, epoch, claimableOP);
+        if (epochNum < _config.maxEpoch) {
```
this should be above OP.transferFrom, keeping a CEI pattern.

0xkarmacoma replied:
in [dcede7a](https://github.com/ethernautdao/ethernaut-op-claim/pull/3/commits/dcede7a0b9eb78f7763e975b22aa517e15855bff)

mariobyn commented on src/OPTokenClaim.sol:
```diff
@@ -54,62 +70,139 @@ contract OPTokenClaim is Ownable {
-        return (expBalance * 5 - 4 ether);
-    }
+            // if total reward of given epoch is greater than 10k OP (MAX_REWARD), reduce reward
+            uint256 factor = 1 * 10 ** 18;
```
maybe use 1 ether instead of 1 * 10**18?

0xkarmacoma replied:
simplified in 2d3841682d14ebc5ca8a33429924c3e9c2ab6d86

0xkarmacoma committed:
Refactor subscribe

Reduce code duplication a little and exit early when resubscribing with no balance change

0xkarmacoma committed:
Increase optimizer_runs

testFuzzClaiming(uint256) (gas: -864 (-0.469%))
testEXPLimit() (gas: -857 (-0.470%))
testClaimingResubscribes() (gas: -1271 (-0.669%))
testReducedReward() (gas: -8699 (-0.694%))
testDoubleClaim() (gas: -1143 (-0.744%))
testExpOwnerCanClaim() (gas: -1566 (-0.794%))
testMaxClaimPeriod() (gas: -3076 (-0.798%))
testResubscribingUpdatesBalance() (gas: -834 (-0.850%))
testNonExpOwnerCanNotClaim() (gas: -221 (-0.940%))
testExtendClaimAuth() (gas: -125 (-0.948%))
testNonExpOwnerCanNotSubscribe() (gas: -207 (-0.974%))
testSubscribeBeforeStart() (gas: -159 (-1.022%))
testExtendClaim() (gas: -169 (-1.073%))
testClaimBeforeStart() (gas: -204 (-1.310%))
Overall gas change: -19395 (-11.756%)

0xkarmacoma committed:
Move resubscription above OP transfer

h/t mario

0xkarmacoma committed:
Optimize storage access in _calcReward

2 SLOADs -> 1 SLOAD

testClaimBeforeStart() (gas: 0 (0.000%))
testExtendClaim() (gas: 0 (0.000%))
testExtendClaimAuth() (gas: 0 (0.000%))
testNonExpOwnerCanNotClaim() (gas: 0 (0.000%))
testNonExpOwnerCanNotSubscribe() (gas: 0 (0.000%))
testResubscribingUpdatesBalance() (gas: 0 (0.000%))
testSubscribeBeforeStart() (gas: 0 (0.000%))
testReducedReward() (gas: -223 (-0.018%))
testFuzzClaiming(uint256) (gas: -223 (-0.122%))
testEXPLimit() (gas: -223 (-0.123%))
testDoubleClaim() (gas: -223 (-0.146%))
testExpOwnerCanClaim() (gas: -446 (-0.228%))
testClaimingResubscribes() (gas: -446 (-0.237%))
testMaxClaimPeriod() (gas: -1071 (-0.280%))
Overall gas change: -2855 (-1.153%)

0xkarmacoma commented on src/OPTokenClaim.sol:
```diff
@@ -54,62 +70,146 @@ contract OPTokenClaim is Ownable {
-            expBalance = 99 ether;
-        }
+        unchecked {
+            uint256 totalReward = totalEXP * 5 - 4 ether * subscribedAccounts;
```
looking closer at the math, I think there's a mistake here 

`(totalEXP * 5) - (4 ether * subscribedAccounts)`

vs 

`(totalEXP * 5 - 4 ether) * subscribedAccounts`

cc @beskay, let's wait to hear from @Dhannte to see if we can simplify the formula. I'd be in favor of getting rid of the `- 4` thing entirely

beskay replied:
The first equation is correct, for example, we have 21 subscribed accounts with 99 EXP each:
totalEXP = 99 * 21 = 2079 EXP
total OP we have to distribute: (99 * 5 - 4) * 21 = 10311 OP

Above equation:
(2079 * 5) - 4 * 21 = 10311 OP

Im fine with simplifying the formula though.

0xkarmacoma committed:
Change award formula

Simplify from

     5 * min(exp, 99) - 4
     (rewards range from 1 to 491 OP)

to

     5 * min(exp, 99)
     (rewards range from 5 to 495 OP)

Plus add more rounding tests

0xkarmacoma committed:
Simplify reward calculation, get rid of numAccounts

0xkarmacoma committed:
Simplify some more

Since we only keep track of the totalEXP per epoch, we can replace the struct and arrays by a simple mapping

Test result: ok. 17 passed; 0 failed; finished in 594.70ms
testExtendClaim() (gas: 0 (0.000%))
testExtendClaimAuth() (gas: 0 (0.000%))
testClaimBeforeStart() (gas: 22 (0.143%))
testSubscribeBeforeStart() (gas: -23 (-0.150%))
testReducedReward99Percent() (gas: -1614167 (-0.758%))
testReducedReward50Percent() (gas: -46367 (-0.852%))
testReduceRewardLimit() (gas: -46367 (-0.852%))
testReducedReward() (gas: -14609 (-1.175%))
testMaxClaimPeriod() (gas: -10163 (-2.672%))
testFuzzClaiming(uint256) (gas: -6338 (-3.467%))
testEXPLimit() (gas: -6338 (-3.506%))
testExpOwnerCanClaim() (gas: -7474 (-3.838%))
testResubscribingUpdatesBalance() (gas: -3685 (-3.889%))
testClaimingResubscribes() (gas: -7474 (-3.983%))
testDoubleClaim() (gas: -6710 (-4.418%))
testNonExpOwnerCanNotSubscribe() (gas: -2428 (-11.521%))
testNonExpOwnerCanNotClaim() (gas: -5417 (-23.248%))
Overall gas change: -1777538 (-64.186%)

beskay reviewed (approved):
looks good!

mariobyn commented on src/OPTokenClaim.sol:
```diff
@@ -9,42 +9,48 @@ contract OPTokenClaim is Ownable {
-    mapping(uint256 => mapping(address => bool)) public epochToAddressClaimed;
+    mapping(uint256 => uint256) public totalEXPAtEpoch;
+
+    mapping(uint256 => mapping(address => uint256)) public epochToSubscribedEXP;
```
this should have a public getter method maybe?

0xkarmacoma replied:
there is the auto-generated public getter method, e.g.:

        // should have subscribed for epoch 5 by now
        assertEq(claimContract.epochToSubscribedEXP(5, alice), 10 ether);

mariobyn commented on src/OPTokenClaim.sol:
```diff
@@ -9,42 +9,48 @@ contract OPTokenClaim is Ownable {
 
-        // TODO: transferOwnership to EthernautDAO multisig?
+        // transfer ownership to multisig
+        _transferOwnership(treasury);
```
a small gas op, use `_treasury`

0xkarmacoma replied:
in 7d4bb60

mariobyn commented on src/OPTokenClaim.sol:
```diff
@@ -54,62 +60,107 @@ contract OPTokenClaim is Ownable {
 
-        uint256 claimableOP = _calcReward(account);
+        uint256 subscribedEXP = epochToSubscribedEXP[epochNum][account];
+        if (subscribedEXP == expBalance) {
```
simply return is weird, i would revert with an error or emit an event at least

0xkarmacoma replied:
in 4840dba

mariobyn commented on src/OPTokenClaim.sol:
```diff
@@ -54,62 +60,107 @@ contract OPTokenClaim is Ownable {
+                reward = 5 * subscribedEXP * MAX_REWARD / totalReward;
+            } else {
+                reward = 5 * subscribedEXP;
+            }
```
if totalReward == 95
user subscribed with 99 EXP
the reward to the user will be 5*99, is that ok?
Can we be in this situation?

0xkarmacoma replied:
do you mean `totalReward == 495`?

if a user subscribed with 99 EXP total reward is 495, that means no other user subscribed. So in that case the user would just get his share normally, nothing weird

mariobyn replied:
total reward can never be less than 495? i did not do the math yet

0xkarmacoma replied:
`totalReward` is `5 * totalEXP`

`totalEXP` is the sum of all `subscribedEXP`, so this should be true for all accounts:

    totalEXP >= subscribedEXP(account)

0xkarmacoma committed:
Use calldata arg instead of stored value

0xkarmacoma committed:
Emit an event in the no-op subscribe case
